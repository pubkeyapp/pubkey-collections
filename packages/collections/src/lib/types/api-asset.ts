export interface ApiAsset {
  id: string
  content: {
    metadata: {
      name: string
      description: string
      attributes: Array<{
        trait_type: string
        value: string
      }>
    }
    files: {
      uri: string
      cdn_uri: string
      mime: string
    }
  }
  grouping: Array<{
    group_type: string
    group_value: string
  }>
}
