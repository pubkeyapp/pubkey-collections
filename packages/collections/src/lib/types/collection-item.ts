import { CollectionTrait } from './collection-trait'

export interface CollectionItem {
  id: string
  name: string
  description?: string
  image?: string
  collectionId: string
  traits: CollectionTrait[]
}
