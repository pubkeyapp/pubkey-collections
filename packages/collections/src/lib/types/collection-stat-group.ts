import { CollectionStat } from './collection-stat'

export interface CollectionStatGroup {
  key: string
  items: CollectionStat[]
}
