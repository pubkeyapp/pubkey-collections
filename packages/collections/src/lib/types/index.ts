export * from './api-asset'
export * from './collection'
export * from './collection-combo'
export * from './collection-item'
export * from './collection-set'
export * from './collection-trait'
export * from './collection-trait-map'
export * from './collection-trait-summary'
