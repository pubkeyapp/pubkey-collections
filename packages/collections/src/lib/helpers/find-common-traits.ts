import { CollectionCombo, CollectionItem, CollectionTrait } from '../types'
import { findItemsWithTraits } from './find-items-with-traits'

function findTraitValues(traits: CollectionTrait[], keys: string[]): string[] {
  return keys.map((key) => traits.find((trait) => trait.key === key)?.value ?? '')
}

function findTraitName(traits: CollectionTrait[]): string {
  return traits.reduce((acc, trait) => {
    const tag = trait.value
    if (acc === '') {
      return tag
    }

    return `${acc}, ${tag}`
  }, '')
}

export function findCommonTraits(
  items: CollectionItem[] = [],
  keys: string[] = [],
  minCommonTraits = 2,
): CollectionCombo[] {
  const result: CollectionCombo[] = []

  for (const item of items) {
    // Get all the other items
    const others = items.filter((other) => other !== item)

    // Check for each of the other items if they have common traits with the current item
    const commonTraits = others
      .map((other) => compareCollectionTraitArrays(keys, item.traits, other.traits))
      .filter((item) => item.length > 1)

    // Loop over the common traits and add them to the result counter
    for (const commonTrait of commonTraits) {
      const name = findTraitName(commonTrait)
      const count = commonTrait.length

      if (!result.find((trait) => trait.name === name && trait.common === count)) {
        result.push({
          name: name,
          common: count,
          traits: commonTrait,
          items: [],
        })
      }
    }
  }

  const foundItems: CollectionItem[] = []

  // Now, loop over the result and add the items to the result
  for (const item of result.sort((a, b) => (b.traits?.length ?? 0) - (a.traits?.length ?? 0))) {
    const found = findItemsWithTraits(items, item.traits).filter((item) => !foundItems.includes(item))

    foundItems.push(...found)
    item.items = found
  }

  return (
    result
      // Filter out the combos that have only one trait
      .filter((item) => (item.traits?.length ?? 0) >= minCommonTraits)
      // Filter out the combos that have only one item
      .filter((item) => (item.items?.length ?? 0) > 1)
  )
}

function compareTwoTraitArrays(a: string[], b: string[]): string[] {
  // The array of common traits
  const result: string[] = []

  // Loop through the first array
  for (const item of a) {
    // If the item is in the second array add it to the result
    result.push(b.includes(item) ? item : '')
  }

  return result
}

function compareCollectionTraitArrays(keys: string[], a: CollectionTrait[], b: CollectionTrait[]): CollectionTrait[] {
  const filteredA = a.filter((trait) => keys.includes(trait.key))
  const filteredB = b.filter((trait) => keys.includes(trait.key))
  const compared = compareTwoTraitArrays(findTraitValues(filteredA, keys), findTraitValues(filteredB, keys))

  return compared
    .map((value, i) => (value !== '' ? filteredA[i] : undefined))
    .filter((trait): trait is CollectionTrait => !!trait)
}
