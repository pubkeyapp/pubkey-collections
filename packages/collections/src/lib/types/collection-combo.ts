import { CollectionItem } from './collection-item'
import { CollectionTrait } from './collection-trait'

export interface CollectionCombo {
  id: string
  name?: string
  links?: Links
  group?: boolean
  sortKey?: string
  common?: number
  traits: CollectionTrait[]
  items?: CollectionItem[]
}

export interface Links {
  homepage?: string
  discord?: string
  twitter?: string
}
