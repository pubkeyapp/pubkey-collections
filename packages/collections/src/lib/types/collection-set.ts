import { Collection } from './collection'
import { CollectionItem } from './collection-item'
import { CollectionTraitMap } from './collection-trait-map'

export interface CollectionSet {
  collection: Collection
  items: CollectionItem[]
  traits: CollectionTraitMap
}
