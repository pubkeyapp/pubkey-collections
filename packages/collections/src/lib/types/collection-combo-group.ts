import { CollectionCombo } from './collection-combo'

export interface CollectionComboGroup {
  name: string
  combos: CollectionCombo[]
}
