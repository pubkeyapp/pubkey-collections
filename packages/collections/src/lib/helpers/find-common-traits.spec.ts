import { CollectionCombo, CollectionItem, CollectionTrait } from '../types'
import { findCommonTraits } from './find-common-traits'

describe('find-common-traits', () => {
  it('should find common traits', () => {
    // Arrange
    const items: CollectionItem[] = [
      {
        id: 'item-1',
        name: 'Item 1',
        traits: [
          { key: 'Background', value: 'Red' },
          { key: 'Foreground', value: 'Green' },
        ],
        collectionId: 'items',
      },
      {
        id: 'item-2',
        name: 'Item 2',
        traits: [
          { key: 'Background', value: 'Blue' },
          { key: 'Foreground', value: 'Yellow' },
        ],
        collectionId: 'items',
      },
      {
        id: 'item-3',
        name: 'Item 3',
        traits: [
          { key: 'Background', value: 'Red' },
          { key: 'Foreground', value: 'White' },
          { key: 'Shape', value: 'Square' },
        ],
        collectionId: 'items',
      },
      {
        id: 'item-4',
        name: 'Item 4',
        traits: [
          { key: 'Background', value: 'Red' },
          { key: 'Foreground', value: 'White' },
          { key: 'Shape', value: 'Circle' },
        ],
        collectionId: 'items',
      },
      {
        id: 'item-5',
        name: 'Item 5',
        traits: [
          { key: 'Background', value: 'Red' },
          { key: 'Foreground', value: 'White' },
          { key: 'Shape', value: 'Triangle' },
        ],
        collectionId: 'items',
      },
      {
        id: 'item-6',
        name: 'Item 6',
        traits: [
          { key: 'Background', value: 'Red' },
          { key: 'Foreground', value: 'White' },
          { key: 'Shape', value: 'Oval' },
          { key: 'Border', value: 'Black' },
        ],
        collectionId: 'items',
      },
      {
        id: 'item-7',
        name: 'Item 7',
        traits: [
          { key: 'Background', value: 'Red' },
          { key: 'Foreground', value: 'White' },
          { key: 'Shape', value: 'Oval' },
          { key: 'Border', value: 'Black' },
        ],
        collectionId: 'items',
      },
    ]
    const keys: string[] = ['Background', 'Foreground', 'Shape', 'Border']

    // Act
    const result = findCommonTraits(items, keys)

    // Assert
    expect(result).toMatchSnapshot()
  })

  it('should work with another example', () => {
    // Arrange
    const items: CollectionItem[] = [
      {
        id: 'id1',
        name: '1',
        collectionId: 'c1',
        traits: [
          { key: 'a', value: '1' },
          { key: 'b', value: '1' },
          { key: 'c', value: '1' },
          { key: 'd', value: 'X' },
        ],
      },
      {
        id: 'id2',
        name: '2',
        collectionId: 'c1',
        traits: [
          { key: 'a', value: '1' },
          { key: 'b', value: '1' },
          { key: 'c', value: '2' },
          { key: 'd', value: 'X' },
        ],
      },
      {
        id: 'id3',
        name: '3',
        collectionId: 'c1',
        traits: [
          { key: 'a', value: '1' },
          { key: 'b', value: '2' },
          { key: 'c', value: '3' },
          { key: 'd', value: 'X' },
        ],
      },
      {
        id: 'id4',
        name: '4',
        collectionId: 'c1',
        traits: [
          { key: 'a', value: '1' },
          { key: 'b', value: '2' },
          { key: 'c', value: '4' },
          { key: 'd', value: 'X' },
        ],
      },
    ]
    // The keys we want to find common traits for
    const keys: string[] = ['a', 'b', 'c']

    // Act
    const result: CollectionCombo[] = findCommonTraits(items, keys)

    // Assert
    expect(result).toMatchSnapshot()
  })
})
