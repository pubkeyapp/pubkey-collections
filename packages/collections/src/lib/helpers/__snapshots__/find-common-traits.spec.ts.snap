// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`find-common-traits should find common traits 1`] = `
[
  {
    "common": 4,
    "id": "red-white-oval-black",
    "items": [
      {
        "collectionId": "items",
        "id": "item-6",
        "name": "Item 6",
        "traits": [
          {
            "key": "Background",
            "value": "Red",
          },
          {
            "key": "Foreground",
            "value": "White",
          },
          {
            "key": "Shape",
            "value": "Oval",
          },
          {
            "key": "Border",
            "value": "Black",
          },
        ],
      },
      {
        "collectionId": "items",
        "id": "item-7",
        "name": "Item 7",
        "traits": [
          {
            "key": "Background",
            "value": "Red",
          },
          {
            "key": "Foreground",
            "value": "White",
          },
          {
            "key": "Shape",
            "value": "Oval",
          },
          {
            "key": "Border",
            "value": "Black",
          },
        ],
      },
    ],
    "traits": [
      {
        "key": "Background",
        "value": "Red",
      },
      {
        "key": "Foreground",
        "value": "White",
      },
      {
        "key": "Shape",
        "value": "Oval",
      },
      {
        "key": "Border",
        "value": "Black",
      },
    ],
  },
  {
    "common": 2,
    "id": "red-white",
    "items": [
      {
        "collectionId": "items",
        "id": "item-3",
        "name": "Item 3",
        "traits": [
          {
            "key": "Background",
            "value": "Red",
          },
          {
            "key": "Foreground",
            "value": "White",
          },
          {
            "key": "Shape",
            "value": "Square",
          },
        ],
      },
      {
        "collectionId": "items",
        "id": "item-4",
        "name": "Item 4",
        "traits": [
          {
            "key": "Background",
            "value": "Red",
          },
          {
            "key": "Foreground",
            "value": "White",
          },
          {
            "key": "Shape",
            "value": "Circle",
          },
        ],
      },
      {
        "collectionId": "items",
        "id": "item-5",
        "name": "Item 5",
        "traits": [
          {
            "key": "Background",
            "value": "Red",
          },
          {
            "key": "Foreground",
            "value": "White",
          },
          {
            "key": "Shape",
            "value": "Triangle",
          },
        ],
      },
    ],
    "traits": [
      {
        "key": "Background",
        "value": "Red",
      },
      {
        "key": "Foreground",
        "value": "White",
      },
    ],
  },
]
`;

exports[`find-common-traits should work with another example 1`] = `
[
  {
    "common": 2,
    "id": "1-2",
    "items": [
      {
        "collectionId": "c1",
        "id": "id3",
        "name": "3",
        "traits": [
          {
            "key": "a",
            "value": "1",
          },
          {
            "key": "b",
            "value": "2",
          },
          {
            "key": "c",
            "value": "3",
          },
          {
            "key": "d",
            "value": "X",
          },
        ],
      },
      {
        "collectionId": "c1",
        "id": "id4",
        "name": "4",
        "traits": [
          {
            "key": "a",
            "value": "1",
          },
          {
            "key": "b",
            "value": "2",
          },
          {
            "key": "c",
            "value": "4",
          },
          {
            "key": "d",
            "value": "X",
          },
        ],
      },
    ],
    "traits": [
      {
        "key": "a",
        "value": "1",
      },
      {
        "key": "b",
        "value": "2",
      },
    ],
  },
]
`;
